### ISSUE TITLE
Whether mstatus.VS should be set to dirty when the vector instructions trigger illegal instruction exception? 

### ISSUE THREAD

**Issue by liweiwei90** [riscv/riscv-isa-sim#1082]

**Comment by aswaterman** [comment]
If VS = 0, that would be illegal. If VS != 0, it is legal but not preferable.

**Comment by liweiwei90** [comment]
I know this check which is done by require_vector_vs. However, the vector instructions have many other checks. What I really mean is whether dirty_vs_state should be called when the other check fails.

**Comment by aswaterman** [comment]
The best option is to avoid dirtying the state if an illegal-instruction exception is raised. But there are some cases where we can make a choice.

**Comment by liweiwei90** [comment]
I find require_vector is often checked before other checks in many cases. So dirtying the state may usually happen when the other check fails.

**Comment by aswaterman** [comment]
That’s technically OK, as long as the state is never dirtied when VS=0. We could choose to make this more precise. Perhaps others should weigh in.

**Comment by liweiwei90** [comment]
OK.

**Comment by scottj97** [comment]
If it's simply a matter of reordering the checks before any state is updated, I'd support that improvement.

**Comment by mlawson-tt** [comment]
One more quasi-related question on this issue: for the floating point instructions, Spike appears to set FS to dirty if an instruction writes to a floating point register (so `fcvt.s.l` sets FS dirty because it writes to an FP register but `fcvt.l.s` does not because it writes to an integer register).  Assuming this behavior is intentional, would it make sense to do the same and only mark VS as dirty if an instruction writes to a vector register (so for example not marking VS dirty if all elements are masked off)?  Or if not, should the FS behavior be changed to match?

**Comment by aswaterman** [comment]
~@mlawson-tt Actually, I think you identified a bug.  Because `fcvt.l.s` writes the FP exception flags, it must mark FS dirty (at least if the flags are actually changed).~

Just kidding.  `fcvt.l.s` does correctly dirty the FP state, inside the `fflags` write routine, if any exceptions are raised.

**Comment by mlawson-tt** [comment]
@aswaterman Ah, I wasn't clear enough there.  `fcvt.l.s` does mark FS as dirty if it produces any exception flags (even if they don't change), but it does not otherwise (whereas `fcvt.s.l` marks FS as dirty regardless as long as it is not illegal)  Spike also marks FS as dirty if on any of the writes to FP CSRs through the `float_csr_t::unlogged_write`, and the same occurs for VS through `vector_csr_t::unlogged_write` and `vxsat_csr_t::unlogged_write`, so I think the saturation case is already covered.

**Comment by aswaterman** [comment]
Right, thanks for the clarification.  Anyway, to answer your question, yes, we have the option of not setting VS to dirty for some vector instructions (e.g. a vector store that is executed with `vstart`=0 and executes to completion, thus keeping `vstart` unchanged, or `vmv.x.s` with `vstart`=0).  Not sure it's worth the effort, but it is an option.

Note that, as implemented, vector instructions explicitly write `vstart` to 0 after completing.  For the cases we'd be trying to filter, that would need to become something like `if (P.VU.vstart->read() != 0) P.VU.vstart->write(0);` to avoid dirtying VS when `vstart` doesn't actually change.

**Comment by mlawson-tt** [comment]
Currently, FS is set dirty if the FP register is written to at all even if the value doesn't change (including the flags), so I think VS could work the same way.  As a result, the `vstart` case would be OK as is because `vstart` is being written to even though the value doesn't change.

Really I think the strongest desire here would just be for FS and VS to behave consistently.  It doesn't really matter which method they use (though I tend to think the current FS behavior of marking as dirty if a register is written to even if the value doesn't change makes sense), but ideally they just both do the same thing in the same scenarios (or at least as close as they can get).  It's a pretty minor issue for the most part, so yeah I'm not sure it's really necessarily worth changing if the current behavior is intentional.  I'm mostly just curious if there is a specific philosophy for when these FS/VS/etc. get marked dirty outside the general requirements in the spec.

Note I haven't looked at it super closely, but it's a possibility that everything works as desired already or is very close because if the instruction is already writing `vstart` at the end, VS will be dirty anyway based on the mark-as-dirty-if-writing-anything behavior.

**Comment by aswaterman** [comment]
@mlawson-tt I agree both with the goal to make FS and VS behave similarly, and with the observation this is almost always the case because of the `vstart` write.

**Comment by mlawson-tt** [comment]
@aswaterman awesome, thanks for the info!
